Add circleci_insights_workflow_run table closes #5

leonid-panich opened the pull request:
Adds table circleci_insights_workflows

@misraved, feedback is welcome! :) 

- "UNION ALL" is needed to query multiple project_slugs. I have no idea how to add a possibility to use "OR" instead.
- I'm not sure how to add a possibility to run parts of UNION ALL in parallel 

# Example query

Get average duration and deployment count of a project for each month
```sql
SELECT 
    w.project_slug, to_char(w.created_at, 'YYYY-MM') as year_month, 
    AVG(w.duration) as average_duration, COUNT(w.id) as deployment_count
FROM 
    circleci_insights_workflow w
WHERE 
    w.workflow_name ='default' AND w.project_slug='gh/companyname/projectname' and w.branch ='main' AND w.status='success'
GROUP BY
    w.project_slug, year_month;
```

cbruno10 commented:
Hey @leonid-panich , sorry for the long wait on the review! @luisffc can chime in with more info, but we're currently looking into understanding more about CircleCI's metrics and their endpoints so we can provide an accurate review on your table. @luisffc will be able to provide his review suggestions and comments some time this week.

Thanks again for creating this new table!

misraved commented on docs/tables/circleci_insights_workflow.md:
````diff
@@ -0,0 +1,19 @@
+### Get average duration and deployment count of a project for each month
+
+```sql
+SELECT 
````
@leonid-panich could you please update the example query to follow the pattern used in other table docs?

To format the query, we recommend using https://www.freeformatter.com/sql-formatter.html#before-output (2 spaces per indent level)

misraved reviewed (commented):
Apologies @leonid-panich for the delayed review and thanks a lot for adding this new table.

I have left some initial comments on the basic structure and the table docs. Please let us know if you have questions 👍.

Also, I think we could add a few more example queries like - `List workflows which are in failed state` and `List workflows created in the last 30 days`.

A follow-up question would be around the `ListConfig` - Is there a way by which we could list all the workflows without the need to pass in a `workflow_name`?

leonid-panich commented:
@misraved 
Thank you for the review! 
I've tried to fix all the issues and have added examples

Unfortunately, I don't know a way to list all the workflows.
It is possible to get insights for all workflows but in a different format:
API documentation: https://circleci.com/docs/api/v2/index.html#operation/getProjectWorkflowMetrics
URL: https://circleci.com/api/v2/insights/{project-slug}/workflows

luisffc commented:
@leonid-panich Thank you for the PR!

I appreciate your contribution. I will delve into the process of retrieving all workflows and provide you with updates on my findings.

Additionally, I noticed that the endpoint includes two additional query parameters: `start-date` and `end-date`. These parameters can be utilized by the plugin for filtering purposes when the `created_at` and/or `stopped_at` field is passed. This approach is similar to how the `branch` field is handled.

Implementing this filtering mechanism minimizes the amount of data transmitted and reduces the processing time required by the plugin.

luisffc commented:
@leonid-panich About loading all workflows, you achieve that by using a Parent Hydrate function.

You can refer to the following example to understand the implementation:
https://github.com/turbot/steampipe-plugin-circleci/blob/main/circleci/table_circleci_context.go#L20
https://github.com/turbot/steampipe-plugin-circleci/blob/main/circleci/table_circleci_context.go#L37

Based on this approach, I recommend creating a new parent hydrate function that performs the following steps:

1. Retrieve all projects, as in: https://github.com/turbot/steampipe-plugin-circleci/blob/main/circleci/table_circleci_project.go#L45

2. For each project, fetch the pipelines, as in: https://github.com/turbot/steampipe-plugin-circleci/blob/main/circleci/table_circleci_pipeline.go#L65

3. For each pipeline, obtain the workflows, as in: https://github.com/turbot/steampipe-plugin-circleci/blob/main/circleci/table_circleci_workflow.go#L55


Additionally, I suggest renaming the table to `circleci_insights_workflow_run` to reflect the type of data returned. Ref: [CircleCI getProjectWorkflowRuns](https://circleci.com/docs/api/v2/index.html#operation/getProjectWorkflowRuns)

luisffc commented on rest/client.go:
```diff
@@ -278,6 +279,59 @@ func (c *Client) ListPipelinesWorkflow(pipelineId string) (*WorkflowResponse, er
 	return workflowResp, nil
 }
 
+func (c *Client) ListInsightsWorkflows(projectSlug, workflowName, branch, startDate, endDate, pageToken string, logger hclog.Logger) (*InsightsWorkflowResponse, error) {
```
oh, sorry, that was my mistake. This is correct, as the plugin also uses `hclog.Logger`

leonid-panich commented:
@luisffc 
I've renamed the table as you suggested and added a ParentHydrate.

I could not find documentation, example implementation and usage of ParentHydrate in the Steampipe docs.
I assumed that the idea is to find all available `project_slugs` and `workflow_names` and 
make these arguments optional for the `circleci_insights_workflow_run` table.

luisffc commented:
@leonid-panich I made some modifications to the `insight_workflow_run` table parent hydration to fix duplicated lines and to avoid hitting rate limit. Let me know if you face any issue
